### PR TITLE
Add virsorter

### PR DESCRIPTION
**Pull request by bernt-matthias**

### CODE REVIEW

File: tools/virsorter/virsorter_run.xml [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+        <requirement type="package" version="3.8.19">python</requirement>
+        <requirement type="package" version="1.1.3">screed</requirement>
+        <requirement type="package" version="3.4">hmmer</requirement>
+        <requirement type="package" version="0.22.1">scikit-learn</requirement>
+        <requirement type="package" version="0.7.0">imbalanced-learn</requirement>
+        <requirement type="package" version="1.2.5">pandas</requirement>
+        <requirement type="package" version="0.13.2">seaborn</requirement>
+        <requirement type="package" version="1.23.5">numpy</requirement>
+        <!-- additional pin (the datamanager failed because of this .. so I'm adding this just be sure) -->
+        <requirement type="package" version="2.7.0">pulp</requirement> <!-- needs to be pinned, because the old-ish snakemake used in vs2 struggles with https://github.com/snakemake/snakemake/issues/2607 -->
```
Comment (wm75) [inline tools/virsorter/virsorter_run.xml:31]: Shouldn't some of this, at least, go upstream into the bioconda recipe to benefit more people?
For my personal taste, take a look at https://github.com/galaxyproject/tools-iuc/blob/main/tools/pangolin/pangolin.xml
There we're only pinning key dependencies that would clearly have a direct impact on reproducibility in Galaxy, but leave the rest to bioconda.

**Review by wm75 — COMMENTED** [review]
Nice, looks like a very interesting tool!

File: data_managers/data_manager_virsorter/data_manager/virsorter_datamanager.xml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+                    <has_text text='"name":"Version @VIRSORTER_DB_VERSION@"'/>
+                </assert_contents>
+            </output>
+        </test>
+    </tests>
+    <help><![CDATA[
+Download and extract virsorter database version @VIRSORTER_DB_VERSION@ from zenodo (https://zenodo.org/records/10594875)
+    ]]></help>
+    <citations>
+        <citation type="doi">10.1186/gb-2014-15-3-r46</citation>
```
Comment (wm75) [inline data_managers/data_manager_virsorter/data_manager/virsorter_datamanager.xml:58]: That seems to be the wrong doi. `10.1186/s40168-020-00990-y` instead?

File: tools/virsorter/virsorter_run.xml (hunk 1 above)
Comment (bernt-matthias) [reply tools/virsorter/virsorter_run.xml:31]: They have just chosen a different way to deploy their software. See https://github.com/jiarong/VirSorter2/issues/205#issuecomment-2181966480. Guess its kind of the typical way snakemake deployments are made, i.e. one conda env for the pipeline and one for the actual tools that are executed. Basically they install https://github.com/jiarong/VirSorter2/blob/master/virsorter/envs/vs2.yaml

But you are right. I can at least remove some that already in bioconda recipe (click and python). Can we losen the pinning for some? Don't know if this works. But I'm pretty sure that we can not remove to many of them.

File: tools/virsorter/virsorter_run.xml [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+            that is used by the snakemake pipeline. in the tool we disable the use of the conda env. 
+            
+            the following are the pinned requirements used in the conda env that would be created by the setup 
+          -->
+        <requirement type="package" version="8.1.7">click</requirement>
+        <requirement type="package" version="1548">last</requirement>
+        <requirement type="package" version="0.3.3">ncbi-genome-download</requirement>
+        <requirement type="package" version="0.16.12">ruamel.yaml</requirement>
+        <requirement type="package" version="2.6.3">prodigal</requirement>
+        <requirement type="package" version="3.8.19">python</requirement>
```
Comment (bernt-matthias) [inline tools/virsorter/virsorter_run.xml:22]:
```suggestion
```

File: tools/virsorter/virsorter_run.xml [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+        <xref type="bio.tools">virsorter</xref>
+    </xrefs>
+    <requirements>
+        <requirement type="package" version="@TOOL_VERSION@">virsorter</requirement>
+        <!-- virsorter (virsorter setup -d db -j 4) creates a conda env yaml (and conda env)
+            that is used by the snakemake pipeline. in the tool we disable the use of the conda env. 
+            
+            the following are the pinned requirements used in the conda env that would be created by the setup 
+          -->
+        <requirement type="package" version="8.1.7">click</requirement>
```
Comment (bernt-matthias) [inline tools/virsorter/virsorter_run.xml:17]:
```suggestion
```

File: tools/virsorter/test-data.sh [hunk 5]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+
+cd test-data/
+wget -O - https://osf.io/v46sc/download | tar -xvz
```
[the remark this answers is not in the document]
Comment (bernt-matthias) [reply tools/virsorter/test-data.sh:4]: Yes, this is the reference data that is used in the test loc file. It's to big for the repo therefore I use this hack.

File: tools/virsorter/virsorter_run.xml [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+
+**Input**
+
+A fasta sequence.
+
+The default score cutoff (0.5) works well known viruses (RefSeq). For the real environmental data, we can expect to get false positives (non-viral) with the default cutoff. Generally, samples with more host (e.g. bulk metaG) and unknown sequences (e.g. soil) tends to have more false positives. We find a score cutoff of 0.9 work well as a cutoff for high confidence hits, but there are also many viral hits with score <0.9. It's difficult to separate the viral and non-viral hits by score alone. So we recommend using the default score cutoff (0.5) for maximal sensitivity and then applying a quality checking step using checkV. Here is a tutorial of [viral identification SOP](https://www.protocols.io/view/viral-sequence-identification-sop-with-virsorter2-btv8nn9w) used in Sullivan Lab.
+
+**Output**
+
+identified viral **sequences**, including two types:
```
[the remark this answers is not in the document]
Comment (bernt-matthias) [reply tools/virsorter/virsorter_run.xml:196]:
```suggestion
identified viral **sequences**, including the following types:
```

File: tools/virsorter/virsorter_run.xml [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+            <when value="--provirus-off">
+                <param argument="--max-orf-per-seq" type="integer" min="-1" value="-1" label="Max number of orf used for computing taxonomic feature" help="" />
+            </when>
+            <when value=""/>
+        </conditional>
+    </inputs>
+    <outputs>
+        <data name="viral_combined" format="fasta" from_work_dir="final-viral-combined.fa" label="${tool.name} on ${on_string}: sequences"/>
+        <data name="viral_score" format="tabular" from_work_dir="final-viral-score.tsv" label="${tool.name} on ${on_string}: scores"/>
+        <data name="viral_boundary" format="tabular" from_work_dir="final-viral-boundary.tsv" label="${tool.name} on ${on_string}: boundary informationsequences">
```
[the remark this answers is not in the document]
Comment (bernt-matthias) [reply tools/virsorter/virsorter_run.xml:96]:
```suggestion
        <data name="viral_boundary" format="tabular" from_work_dir="final-viral-boundary.tsv" label="${tool.name} on ${on_string}: boundary information">
```

File: tools/virsorter/virsorter_run.xml [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+        <param argument="--seqname-suffix-off" type="boolean" truevalue="--" falsevalue="" checked="false" label="Do not add suffix to sequence names" help="By default suffix:  ||full, ||{i}_partial, ||lt2gene is appended. ote that this might cause partial seqs from the same contig to have the same name; this option is could be used when you are sure there is one partial sequence at max from each contig" />
+        <param argument="--prep-for-dramv" type="boolean" truevalue="--prep-for-dramv" falsevalue="" checked="false" label="Generate viral seqfile and viral-affi-contigs for DRAMv" />
+        <conditional name="provirus_cond">
+            <param argument="--provirus-off" type="select" label="Extract provirus after classifying full contigs" help="">
+                <option value="--provirus-off">No provirus extraction</option>
+                <option value="" selected="true">provirus extraction</option>
+            </param>
+            <when value="--provirus-off">
+                <param argument="--max-orf-per-seq" type="integer" min="-1" value="-1" label="Max number of orf used for computing taxonomic feature" help="" />
```
Comment (bernt-matthias) [inline tools/virsorter/virsorter_run.xml:88]:
```suggestion
                <param argument="--max-orf-per-seq" type="integer" min="-1" value="-1" label="Max number of orf used for computing taxonomic feature" help=" This option subsamples ORFs if a sequence has more ORFs than the number provided" />
```

File: data_managers/data_manager_virsorter/data_manager/virsorter_datamanager.xml (hunk 2 above)
Comment (bernt-matthias) [reply data_managers/data_manager_virsorter/data_manager/virsorter_datamanager.xml:58]:
```suggestion
        <citation type="doi">10.1186/s40168-020-00990-y</citation>
```

**Comment by bernt-matthias** [comment]
Thanks @wm75 .. there are a few more similar tools on my list.

**Pull request merged**